gluten-core/src/main/scala/io/glutenproject/extension/columnar/ColumnarPullOutProject.scala

```diff
@@ -0,0 +1,279 @@
+    plan.setTagValue(TAG, hint)
+  }
+}
+
+trait ColumnarPullOutProject {
+  val applyLocally: PartialFunction[SparkPlan, SparkPlan]
+
+  def getPulledOutPlan[T <: SparkPlan](plan: SparkPlan): T = {
+    // Use applyLocally to only transform current plan, not transform its children.
+    val pulledOutPlan = applyLocally.lift(plan).getOrElse(plan)
+    pulledOutPlan match {
+      case p: ProjectExec =>
+        p.child.asInstanceOf[T]
+      case other =>
+        other.asInstanceOf[T]
+    }
+  }
+}
+
+object ColumnarPullOutProject extends Rule[SparkPlan] {
+
+  private lazy val validationRules: List[Rule[SparkPlan] with ColumnarPullOutProject] = List(
+    ColumnarPullOutPreProject(validation = true),
+    ColumnarPullOutPostProject(validation = true)
+  )
+
+  private lazy val rules: List[Rule[SparkPlan]] = List(
+    ColumnarPullOutPreProject(),
+    ColumnarPullOutPostProject()
+  )
+
+  def applyPullOutColumnarPreRules(plan: SparkPlan): SparkPlan = {
+    rules.foldLeft(plan) { case (latestPlan, rule) => rule.apply(latestPlan) }
+  }
+
+  def getPulledOutPlanLocally[T <: SparkPlan](plan: SparkPlan): T = {
+    validationRules
+      .foldLeft(plan) {
+        case (latestPlan, rule) =>
+          rule.getPulledOutPlan[T](latestPlan)
+      }
+      .asInstanceOf[T]
+  }
+
+  def applyPullOutGlutenPlanRules(plan: GlutenPlan): SparkPlan =
+    GlutenPlanPullOutProject.apply(plan)
+
+  override def apply(plan: SparkPlan): SparkPlan = applyPullOutColumnarPreRules(plan)
+}
+
+case class ColumnarPullOutPostProject(validation: Boolean = false)
+  extends Rule[SparkPlan]
+  with ColumnarPullOutProject
+  with PullOutProjectHelper
+  with AliasHelper {
+
+  private def needsPostProjection(plan: SparkPlan): Boolean = {
+    // For validation, no need to check the TransformHint in plan.
+    if (!validation && !supportTransform(plan)) {
+      return false
+    }
+    plan match {
+      case agg: BaseAggregateExec =>
+        val pullOutHelper = HashAggregateExecPullOutHelper(agg)
+        val allAggregateResultAttributes = pullOutHelper.allAggregateResultAttributes
+        // If the result expressions has different size with output attribute,
+        // post-projection is needed.
+        agg.resultExpressions.size != allAggregateResultAttributes.size ||
+        // Compare each item in result expressions and output attributes. Attribute in Alias
+        // should be trimmed before checking.
+        agg.resultExpressions.map(trimAliases).zip(allAggregateResultAttributes).exists {
+          case (exprAttr: Attribute, resAttr) =>
+            // If the result attribute and result expression has different name or type,
+            // post-projection is needed.
+            exprAttr.name != resAttr.name || exprAttr.dataType != resAttr.dataType
+          case _ =>
+            // If result expression is not instance of Attribute,
+            // post-projection is needed.
+            true
+        }
+      case _ => false
```

Comment:
Do we need the sort check here?

Reply:
Sort is different from Agg, if it has pre-project, it always needs a post-project, so I pullout pre and post project together in `ColumnarPullOutPreProject` for `SortExec`.